src/v/cloud_storage/partition_manifest.cc

```diff
@@ -856,6 +856,21 @@ void partition_manifest::reset_scrubbing_metadata() {
 
 std::optional<size_t> partition_manifest::move_aligned_offset_range(
   const segment_meta& replacing_segment) {
+    // nullopt means that the replacing_segment can't be inserted in the
+    // manifest, either by breaking an invariant in segment_map, or by failing
+    // to generate a distinct name
+    if (unlikely(!_segments.can_be_inserted(replacing_segment))) {
+        // log an error, as it's not expected for a segment at this stage to
+        // fail this check. safe_segment_meta_to_add() fail before this
+        vlog(
+          cst_log.error,
+          "{} segment {:s} can't be inserted because it's covered by another "
+          "in the manifest",
+          _ntp,
+          replacing_segment);
+        return std::nullopt;
```

Comment:
how do callers react to this? the way this is described is that an invariant, but we signal to the caller about the condition.

Reply:
the caller recognizes that the segment_meta has not been inserted, so it's not allowed to delete the segment file. this check should be already implemented in partition_manifest::safe_segment_meta_to_add() https://github.com/redpanda-data/redpanda/blob/0df54191a8199c21708b698e97faa95efd48e465/src/v/cloud_storage/partition_manifest.cc#L1036-L1074 but i don't see the exact check.

the unit test that i wrote checks for this, but maybe i can ensure it and see what breaks in ducktape

Reply:
Let's see
It's an error case, but if the range is already in the manifest, it means that the smaller range that we are trying to add is old data that has been superseded by the stuff in the manifest. 
It shouldn't be a data loss event. 
On the other hand, it is a case that should never happen so a vassert is a stronger signal.
Maybe we can have a vassert here for this testing period and schedule a PR to remove it a couple of weeks before the complete release? I worry that it could cause a crash loop _if_ it happens in the wild